drivers/pwm/pwm_nrfx.c

```diff
@@ -12,86 +12,245 @@ LOG_MODULE_REGISTER(pwm_nrfx);
 
 #define PWM_NRFX_CH_VALUE_NORMAL        (1UL << 15)
 #define PWM_NRFX_CH_VALUE_INVERTED      (0)
+#define NRFX_PWM_PIN_INVERTED_MASK      0x80
+#define NRFX_PWM_PIN_MASK               0x7F
+#define NRFX_PWM_POLARITY_MSK           0x8000
+#define NRFX_PWM_COMPARE_MASK           0x7FFF
 
 struct pwm_nrfx_config {
 	nrfx_pwm_t pwm;
-	nrfx_pwm_config_t config;
-	nrf_pwm_sequence_t seq;
 };
 
 struct pwm_nrfx_data {
+	nrfx_pwm_config_t config;
+	nrf_pwm_sequence_t seq;
 	u16_t current[NRF_PWM_CHANNEL_COUNT];
-	u16_t top_value;
+	u16_t pwm_instance_in_use;
+	u32_t period_cycles;
+	u16_t prescaler_divisor;
 };
 
+static u32_t pwm_period_check_and_set(
+	struct pwm_nrfx_data *pdata, NRF_PWM_Type *pInstance,
+	u32_t pwm, u32_t period_cycles)
+{
+	/* Fail if requested period does not match already running period */
+	if (pdata->pwm_instance_in_use &&
+			(period_cycles != pdata->period_cycles)) {
+		LOG_ERR("Fail:requested period cycles:%d, != used %d\n",
+			period_cycles, pdata->period_cycles);
+		return -EINVAL;
+	}
+
+	/* Check if period_cycles is above COUNTERTOP MAX value, if so, we
+	 * have to see if we can change frequency to something that will fit
+	 */
+	if (period_cycles > PWM_COUNTERTOP_COUNTERTOP_Msk) {
+		/* See if there is a prescaler that will make it work: */
+		bool matching_prescaler_found = false;
+
+		/* Go through all available prescaler values on device.
+		 * nRF52832 has 0-7 (Div1 - Div128)
+		 */
+		for (int i = 0; i <= PWM_PRESCALER_PRESCALER_Msk; i++) {
+			u16_t divisor = 1 << i;
+			u32_t new_countertop = period_cycles / divisor;
+
+			/* If we find value that fits, set it, continue */
+			if (new_countertop <= PWM_COUNTERTOP_COUNTERTOP_Msk) {
+				pdata->config.base_clock = i;
+				pdata->config.top_value = new_countertop;
+				pdata->prescaler_divisor = divisor;
+				pdata->period_cycles = period_cycles;
+				matching_prescaler_found = true;
+				break;
+			}
+		}
+
+		/* Check if able to find matching prescaler and countertop */
+		if (matching_prescaler_found == false) {
+			LOG_ERR(
+			"Matching prescaler for period_cycles %d not found.\n",
+			period_cycles);
```

Comment:
Such indentation is misleading. Consider shortening the error message.

Reply:
Done.